src/data/list/duplicate.lean

```diff
@@ -0,0 +1,227 @@
+/-
+Copyright (c) 2021 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky, Chris Hughes
+-/
+import data.list.basic
+import data.fin
+
+/-!
+# List duplicates
+
+## Main definitions
+
+* `list.duplicate x l : Prop` is an inductive property that holds when `x` is a duplicate in `l`
+
+-/
+
+variable {α : Type*}
+
+namespace list
+
+/-- Property that an element `x : α` of `l : list α` can be found in the list more than once. -/
+inductive duplicate (x : α) : list α → Prop
+| cons_mem {l : list α} : x ∈ l → duplicate (x :: l)
+| cons_duplicate {y : α} {l : list α} : duplicate l → duplicate (y :: l)
+
+local infix ` ∈+ `:50 := list.duplicate
+
+variables {l : list α} {x : α}
+
+lemma mem.duplicate_cons_self (h : x ∈ l) : x ∈+ x :: l := duplicate.cons_mem h
+
+lemma duplicate.duplicate_cons (h : x ∈+ l) (y : α) : x ∈+ y :: l := duplicate.cons_duplicate h
+
+lemma duplicate.mem (h : x ∈+ l) : x ∈ l :=
+begin
+  induction h with l' h y l' h hm,
+  { exact mem_cons_self _ _ },
+  { exact mem_cons_of_mem _ hm }
+end
+
+lemma duplicate.mem_cons_self (h : x ∈+ x :: l) : x ∈ l :=
+begin
+  cases h with _ h _ _ h,
+  { exact h },
+  { exact h.mem }
+end
+
+@[simp] lemma duplicate_cons_self_iff : x ∈+ x :: l ↔ x ∈ l :=
+⟨duplicate.mem_cons_self, mem.duplicate_cons_self⟩
+
+lemma duplicate.ne_nil (h : x ∈+ l) : l ≠ [] :=
+λ H, (mem_nil_iff x).mp (H ▸ h.mem)
+
+@[simp] lemma not_duplicate_nil (x : α) : ¬ x ∈+ [] :=
+λ H, by simpa using H.ne_nil
+
+lemma duplicate.ne_singleton (h : x ∈+ l) (y : α) : l ≠ [y] :=
+begin
+  induction h with l' h z l' h hm,
+  { simp [ne_nil_of_mem h] },
+  { simp [ne_nil_of_mem h.mem] }
+end
+
+@[simp] lemma not_duplicate_singleton (x y : α) : ¬ x ∈+ [y] :=
+λ H, by simpa using H.ne_singleton _
+
+lemma duplicate_cons_iff {y : α} : x ∈+ y :: l ↔ (y = x ∧ x ∈ l) ∨ x ∈+ l :=
+begin
+  refine ⟨λ h, _, λ h, _⟩,
+  { cases h with _ hm _ _ hm,
+    { exact or.inl ⟨rfl, hm⟩ },
+    { exact or.inr hm } },
+  { rcases h with ⟨rfl|h⟩|h,
+    { simpa },
+    { exact h.duplicate_cons _ } }
+end
+
+lemma duplicate.of_duplicate_cons {y : α} (h : x ∈+ y :: l) (hx : x ≠ y) : x ∈+ l :=
+by simpa [duplicate_cons_iff, hx.symm] using h
+
+lemma duplicate_cons_iff_of_ne {y : α} (hne : x ≠ y) : x ∈+ y :: l ↔ x ∈+ l :=
+by simp [duplicate_cons_iff, hne.symm]
+
+lemma duplicate.mono_sublist {l' : list α} (hx : x ∈+ l) (h : l <+ l') : x ∈+ l' :=
+begin
+  induction h with l₁ l₂ y h IH l₁ l₂ y h IH,
+  { simpa using hx.ne_nil },
+  { exact (IH hx).duplicate_cons _ },
+  { rw duplicate_cons_iff at hx ⊢,
+    rcases hx with ⟨rfl, hx⟩|hx,
+    { simp [h.subset hx] },
+    { simp [IH hx] } }
+end
+
+/-- The contrapositive of `list.nodup_iff_sublist`. -/
```

Comment:
Do you want to add lemmas for the relation between `nodup` and `∈+`? For instance, `l.nodup ↔ ∀ x, ¬(x ∈+ l)` - I expect they follow easily from this or `duplicate_iff_count_le_two` but I think it'd still be nice to have lemmas to cleanly transfer between `duplicate` and `nodup`

Reply:
I did not add these lemmas initially because I was trying to keep imports minimal. Now that this file imports `data.fin`, then `nodup` is imported transitively, so I can add them.